File: ports/lua/test/skin_tone_variation_forgery_test.lua
local h = require("test_helper")
local policy = require("unicode_lua.security.policy")
local calculus = require("unicode_lua.security.calculus")
local stvf = require("unicode_lua.security.identity.skin_tone_variation_forgery")

local Family = calculus.Family

local function tag(input)
  return stvf.classification_tag(stvf.detect(input))
end

-- ── Shared context-free fixture ─────────────────────────────────────────
-- Drive every case through the detector and map the classification to the
-- family reason code, exactly as the policy layer would.

local fixture = h.json_file("testdata/fixtures/security/detectors/skin_tone_variation_forgery.json")
local fixture_cases = 0
for _, case in ipairs(fixture.cases) do
  fixture_cases = fixture_cases + 1
  local verdict = stvf.detect(case.input)
  local t = stvf.classification_tag(verdict)
  local codes = {}
  if t ~= nil then
    codes[#codes + 1] = policy.reason_code(Family.SkinToneVariationForgery, t)
  end
  for _, required in ipairs(case.required_findings) do
    h.assert_includes(codes, required, "fixture/" .. case.name)
  end
  if #case.required_findings == 0 and t ~= nil then
    error("fixture/" .. case.name .. " unexpected finding " .. t)
  end
end

-- ── §6 detect spot checks (one per Rust theorem) ─────────────────────────

-- detect_empty_clear
local vempty = stvf.detect({})
h.assert_equal(stvf.is_clear(vempty), true, "empty is clear")
h.assert_equal(stvf.classification_tag(vempty), nil, "empty tag nil")

-- detect_ascii_clear — "He"
h.assert_equal(stvf.is_clear(stvf.detect({ 0x48, 0x65 })), true, "ascii clear")

-- detect_plain_emoji_clear — grinning face
h.assert_equal(stvf.is_clear(stvf.detect({ 0x1F600 })), true, "plain emoji clear")

-- detect_wave_skin_tone_clear — waving hand (a modifier base) + one skin tone
local vwave = stvf.detect({ 0x1F44B, 0x1F3FB })
h.assert_equal(stvf.is_clear(vwave), true, "wave+single-tone clear")
h.assert_equal(vwave.skin_tone_count, 1, "wave skin_tone_count")

-- detect_stacked_skin_tones — waving hand + two skin tones
local vstack = stvf.detect({ 0x1F44B, 0x1F3FB, 0x1F3FC })
h.assert_equal(tag({ 0x1F44B, 0x1F3FB, 0x1F3FC }), "StackedSkinTones", "stacked tag")
h.assert_equal(vstack.classify.positions[1], 1, "stacked positions[1]")
h.assert_equal(vstack.classify.positions[2], 2, "stacked positions[2]")

-- detect_invalid_target_ascii — skin tone on ASCII 'A'
local vinva = stvf.detect({ 0x0041, 0x1F3FB })
h.assert_equal(stvf.classification_tag(vinva), "InvalidSkinToneTarget", "invalid-ascii tag")
h.assert_equal(vinva.classify.positions[1], 1, "invalid-ascii positions[1]")

-- detect_invalid_target_smiley — skin tone on grinning face (not a modifier base)
h.assert_equal(tag({ 0x1F600, 0x1F3FB }), "InvalidSkinToneTarget", "invalid-smiley tag")

-- detect_forced_text_style — VS15 on grinning face (Emoji_Presentation)
local vforced = stvf.detect({ 0x1F600, 0xFE0E })
h.assert_equal(stvf.classification_tag(vforced), "ForcedTextStyle", "forced-text-style tag")
h.assert_equal(vforced.variation_selector15_count, 1, "vs15 count")

-- Reused-predicate sanity: skin-tone predicate, modifier-base, presentation.
h.assert_equal(stvf.is_skin_tone(0x1F3FB), true, "1F3FB is skin tone")
h.assert_equal(stvf.is_skin_tone_base(0x1F44B), true, "1F44B is modifier base")
h.assert_equal(stvf.is_emoji_presentation(0x1F600), true, "1F600 is emoji presentation")
h.assert_equal(stvf.is_skin_tone_base(0x1F600), false, "1F600 not a modifier base")

-- Reason code composition.
h.assert_equal(
  policy.reason_code(Family.SkinToneVariationForgery, "StackedSkinTones"),
  "unicode.security.I.skin-tone-variation-forgery.StackedSkinTones",
  "reason code")

print("skin_tone_variation_forgery: fixture cases = " .. fixture_cases)

File: ports/php/test/skin_tone_variation_forgery_test.php
<?php

declare(strict_types=1);

require_once __DIR__ . '/test_helper.php';

use UnicodePhp\Security\Identity\SkinToneVariationForgery;
use UnicodePhp\Security\Mode;
use UnicodePhp\Security\Policy;
use UnicodePhp\Security\Profile;

/** @param list<int> $input */
function skin_tone_variation_forgery_tag(array $input): ?string
{
    return SkinToneVariationForgery::detect($input)->classify->tag();
}

// ── (a) shared context-free fixture, driven through scanSkinToneVariationForgery ──
$fixture = fixture_json('detectors/skin_tone_variation_forgery.json');
$fixtureCases = 0;
foreach ($fixture['cases'] as $case) {
    $verdict = Policy::scanSkinToneVariationForgery(Profile::GatewayHeader, Mode::Observe, $case['input']);
    $codes = verdict_codes($verdict);
    foreach ($case['required_findings'] as $required) {
        assert_includes_value($codes, $required, 'fixture/' . $case['name']);
    }
    if ($case['required_findings'] === []) {
        $needle = '.' . $fixture['family'] . '.';
        foreach ($codes as $code) {
            if (str_contains($code, $needle)) {
                throw new RuntimeException('fixture/' . $case['name'] . ' unexpected family ' . $fixture['family']);
            }
        }
    }
    $fixtureCases++;
}

// ── data-layer sanity: confirm reuse of the port's own predicates ────────────
// Skin-tone modifier — EmojiZwjIntegrity::isEmojiModifier (U+1F3FB..U+1F3FF).
assert_same_value(true, SkinToneVariationForgery::isSkinTone(0x1F3FB), 'skin tone light');
assert_same_value(true, SkinToneVariationForgery::isSkinTone(0x1F3FF), 'skin tone dark');
assert_same_value(false, SkinToneVariationForgery::isSkinTone(0x1F44B), 'wave not skin tone');
// Emoji_Modifier_Base — parsed from the bundled emoji-data.txt.
assert_same_value(true, SkinToneVariationForgery::isSkinToneBase(0x1F44B), 'waving hand is modifier base');
assert_same_value(false, SkinToneVariationForgery::isSkinToneBase(0x0041), 'ascii A not modifier base');
assert_same_value(false, SkinToneVariationForgery::isSkinToneBase(0x1F600), 'grinning face not modifier base');
// Emoji_Presentation — parsed from the bundled emoji-data.txt.
assert_same_value(true, SkinToneVariationForgery::isEmojiPresentation(0x1F600), 'grinning face is emoji presentation');
assert_same_value(false, SkinToneVariationForgery::isEmojiPresentation(0x0041), 'ascii A not emoji presentation');
// Variation selectors.
assert_same_value(true, SkinToneVariationForgery::isVs15(0xFE0E), 'vs15');
assert_same_value(false, SkinToneVariationForgery::isVs15(0xFE0F), 'vs16 is not vs15');
assert_same_value(true, SkinToneVariationForgery::isVs16(0xFE0F), 'vs16');
assert_same_value(false, SkinToneVariationForgery::isVs16(0xFE0E), 'vs15 is not vs16');

// ── §5 detect spot checks (one per rust `detect_*` test) ─────────────────────

// detect_empty_clear
assert_same_value(true, SkinToneVariationForgery::detect([])->classify->isClear(), 'empty clear');

// detect_ascii_clear — "He"
assert_same_value(true, SkinToneVariationForgery::detect([0x48, 0x65])->classify->isClear(), 'ascii clear');

// detect_plain_emoji_clear — grinning face
assert_same_value(true, SkinToneVariationForgery::detect([0x1F600])->classify->isClear(), 'plain emoji clear');

// detect_wave_skin_tone_clear — waving hand (a modifier base) + one skin tone.
$v = SkinToneVariationForgery::detect([0x1F44B, 0x1F3FB]);
assert_same_value(true, $v->classify->isClear(), 'wave skin tone clear');
assert_same_value(1, $v->skinToneCount, 'wave skin tone count');

// detect_stacked_skin_tones — waving hand + two skin tones.
$v = SkinToneVariationForgery::detect([0x1F44B, 0x1F3FB, 0x1F3FC]);
assert_same_value('StackedSkinTones', $v->classify->tag(), 'stacked skin tones tag');
assert_same_value([1, 2], $v->classify->positions(), 'stacked skin tones positions');

// detect_invalid_target_ascii — skin tone on ASCII 'A'.
$v = SkinToneVariationForgery::detect([0x0041, 0x1F3FB]);
assert_same_value('InvalidSkinToneTarget', $v->classify->tag(), 'invalid target ascii tag');
assert_same_value([1], $v->classify->positions(), 'invalid target ascii positions');

// detect_invalid_target_smiley — skin tone on grinning face (not a modifier base).
assert_same_value('InvalidSkinToneTarget', skin_tone_variation_forgery_tag([0x1F600, 0x1F3FB]), 'invalid target smiley');

// detect_forced_text_style — VS15 on grinning face (Emoji_Presentation).
$v = SkinToneVariationForgery::detect([0x1F600, 0xFE0E]);
assert_same_value('ForcedTextStyle', $v->classify->tag(), 'forced text style tag');
assert_same_value(1, $v->variationSelector15Count, 'forced text style vs15 count');

// ── priority-ladder structural check ─────────────────────────────────────────

// Stacked skin tones outrank the invalid-target case they also satisfy.
assert_same_value('StackedSkinTones', skin_tone_variation_forgery_tag([0x0041, 0x1F3FB, 0x1F3FC]), 'stacked beats invalid target');

// ── reason-code stability ────────────────────────────────────────────────────
assert_same_value(
    'unicode.security.I.skin-tone-variation-forgery.StackedSkinTones',
    Policy::reasonCode(\UnicodePhp\Security\Family::SkinToneVariationForgery, 'StackedSkinTones'),
    'reason code stacked',
);
assert_same_value(
    'unicode.security.I.skin-tone-variation-forgery.ForcedTextStyle',
    Policy::reasonCode(\UnicodePhp\Security\Family::SkinToneVariationForgery, 'ForcedTextStyle'),
    'reason code forced text style',
);

echo 'skin_tone_variation_forgery_test: OK (' . $fixtureCases . " fixture cases)\n";
